src/qat/ir/instructions.py

```diff
@@ -0,0 +1,1032 @@
+        """If you need a name, use generate_name and pass in existing values."""
+        super().__init__(name=name)
+
+    @staticmethod
+    def with_random_name(existing_names=None):
+        return Label(Label.generate_name(existing_names))
+
+    @staticmethod
+    def generate_name(existing_names=None):
+        return build_generated_name(existing_names)
+
+    def __repr__(self):
+        return f"{self.name}:"
+
+
+class Jump(Instruction):
+    """
+    Classic jump instruction, should be linked to label with an optional condition.
+    """
+
+    inst: Literal["Jump"] = "Jump"
+    target: str
+    condition: Any
+
+    def __init__(self, target: Union[str, Label], condition=None, **kwargs):
+        return super().__init__(
+            target=target.name if isinstance(target, Label) else target,
+            condition=condition,
+        )
+
+    def __repr__(self):
+        if self.condition is not None:
+            return f"if {str(self.condition)} -> {str(self.target)}"
+        else:
+            return f"-> {str(self.target)}"
+
+
+class ResultsProcessing(Instruction):
+    """
+    A meta-instruction that stores how the results are processed.
+    """
+
+    inst: Literal["ResultsProcessing"] = "ResultsProcessing"
+    variable: str
+    res_processing: InlineResultsProcessing
+
+    def __init__(self, variable: str, res_processing: InlineResultsProcessing, **kwargs):
+        super().__init__(variable=variable, res_processing=res_processing)
+
+    def __repr__(self):
+        return f"{self.variable}: {str(self.res_processing.name)}"
+
+
+### Quantum Instructions
+class QuantumInstruction(Instruction):
+    """
+    Any node that deals particularly with quantum operations. All quantum operations
+    must have some sort of target on the quantum computer, such as a qubit, channel, or
+    another form of component.
+    """
+
+    inst: Literal["QuantumInstruction"] = "QuantumInstruction"
+    quantum_targets: Union[set[str], str]
+
+    @property
+    def duration(self):
+        return 0.0
+
+    @field_validator("quantum_targets", mode="before")
+    @classmethod
+    def _components_to_ids(cls, targets):
+        """
+        Fetches the IDs from quantum components.
+        """
+        if isinstance(targets, (set, list, tuple)):
+            targets = set(
+                [
+                    itm.full_id() if isinstance(itm, QuantumComponent) else itm
+                    for itm in targets
+                ]
+            )
+            targets = [val for val in targets][0] if len(targets) == 1 else targets
+        else:
+            targets = (
+                targets.full_id() if isinstance(targets, QuantumComponent) else targets
+            )
+        return targets
+
+
+class PhaseShift(QuantumInstruction):
+    """
+    A PhaseShift instruction is used to change the phase of waveforms sent down
+    the pulse channel.
+    """
+
+    inst: Literal["PhaseShift"] = "PhaseShift"
+    quantum_targets: str
```

Comment:
Why can't we make the type of this variable a `PulseChannel` or a collection thereof?

Reply:
So with the legacy hardware models, since they're not derived on the `BaseModel`, this wouldn't work.

Of course with the new hardware models this isn't an issue. However, when we serialize, the details of the hardware model (e.g. pulse channels) would also be serialized. If we really wanted the instruction to store the entire pulse channel, we could probably do something with `field_serializers` and `field_validators` to deal with this.

My thoughts are it is best to separate this entirely (just storing ids/uuids). There is the obvious downside that we would need to retrieve the pulse channels at compilation. However, the `TriagePass` bundles instructions by pulse channels so this should only need to be done once per pulse channel.